src/cmd/flux-exec.c

```diff
@@ -226,18 +287,28 @@ static void stdin_cb (flux_reactor_t *r,
     struct fbuf *fb = fbuf_read_watcher_get_buffer (w);
     flux_subprocess_t *p;
     const char *ptr;
-    int lenp;
+    int len, lenp;
+    int min_credits = subprocess_min_credits ();
 
-    if (!(ptr = fbuf_read (fb, -1, &lenp)))
+    if (!(ptr = fbuf_read (fb, min_credits, &lenp)))
         log_err_exit ("fbuf_read");
 
     if (lenp) {
         p = zlistx_first (subprocesses);
         while (p) {
             if (flux_subprocess_state (p) == FLUX_SUBPROCESS_INIT
                 || flux_subprocess_state (p) == FLUX_SUBPROCESS_RUNNING) {
-                if (flux_subprocess_write (p, "stdin", ptr, lenp) < 0)
+                int updated_credits;
+                if ((len = flux_subprocess_write (p, "stdin", ptr, lenp)) < 0)
                     log_err_exit ("flux_subprocess_write");
+                /* N.B. since we are subtracting the same number of credits
+                 * from all subprocesses, the sorted order in the credits list
+                 * should not change
+                 */
+                updated_credits = subprocess_update_credits (p, -1*len, false);
+                /* if one subprocess has no more credits, stop stdin watcher */
+                if (updated_credits == 0)
+                    flux_watcher_stop (stdin_w);
```

Comment:
I would just not have `subprocess_update_credits()` return anything and make a call to `subprocess_min_credits()` once after the loop and stop the watcher there if zero, since that's now a cheapish operation.